ethpm/tools/builder.py

```diff
@@ -0,0 +1,332 @@
+    """
+    return _links(link_dict)
+
+
+@curry
+def _links(link_dict: Dict[str, str], manifest: Manifest) -> Manifest:
+    return assoc_in(manifest, ("meta", "links"), link_dict)
+
+
+#
+# Sources
+#
+
+
+def get_names_and_paths(compiler_output: Dict[str, Any]) -> Dict[str, str]:
+    """
+    Return a mapping of contract name to relative path as defined in compiler output.
+    """
+    return {
+        name: path
+        for path, sep, name in map(
+            operator.methodcaller("partition", ":"), compiler_output
+        )
+    }
+
+
+def inline_source(
+    name: str, compiler_output: Dict[str, Any], package_root_dir: str = None
+) -> Manifest:
+    """
+    Return a copy of manifest with added field to
+    "sources": {relative_source_path: contract_source_data}.
+
+    If `package_root_dir` is not provided, cwd is expected to resolve the relative
+    path to the source as defined in the compiler output.
+    """
+    return _inline_source(name, compiler_output, package_root_dir)
+
+
+@curry
+def _inline_source(
+    name: str,
+    compiler_output: Dict[str, Any],
+    package_root_dir: str,
+    manifest: Manifest,
+) -> Manifest:
+    names_and_paths = get_names_and_paths(compiler_output)
+    cwd = Path(os.getcwd())
+    source_path_suffix = names_and_paths[name]
+
+    if (cwd / source_path_suffix).is_file():
+        source_data = (cwd / source_path_suffix).read_text()
+    elif package_root_dir and (package_root_dir / source_path_suffix).is_file():
+        source_data = (package_root_dir / source_path_suffix).read_text()
+    else:
+        raise ManifestBuildingError(
+            "Contract source cannot be resolved, please make sure that the working "
+            "directory is set to the correct directory or provide `pacakge_root_dir`."
+        )
+
+    if "sources" not in manifest:
+        manifest["sources"] = {}
+
+    return assoc(
+        manifest,
+        "sources",
+        assoc(manifest["sources"], "./{0}".format(source_path_suffix), source_data),
+    )
+
+
+def pin_source(
+    name: str,
+    compiler_output: Dict[str, Any],
+    ipfs_backend: BaseIPFSBackend,
+    package_root_dir: str = None,
+) -> Manifest:
+    """
+    Pins source to IPFS and returns a copy of manifest with added field to
+    "sources": {relative_source_path: IFPS URI}.
+
+    If `package_root_dir` is not provided, cwd is expected to resolve the relative path
+    to the source as defined in the compiler output.
+    """
+    return _pin_source(name, compiler_output, ipfs_backend, package_root_dir)
+
+
+@curry
+def _pin_source(
+    name: str,
+    compiler_output: Dict[str, Any],
+    ipfs_backend: BaseIPFSBackend,
+    package_root_dir: str,
+    manifest: Manifest,
+) -> Manifest:
+    names_and_paths = get_names_and_paths(compiler_output)
+    source_path = names_and_paths[name]
+    if package_root_dir:
+        (ipfs_data,) = ipfs_backend.pin_assets(package_root_dir / source_path)
```

Comment:
This seems like something worth running pre-validation on to check if the file is present and to throw a friendly error here where you have context of what is happening rather than the backend throwing a potentially harder to understand exception when it can't find the file.